pkg/expr/nodes.go

```diff
@@ -191,69 +192,100 @@ func (s *Service) buildDSNode(dp *simple.DirectedGraph, rn *rawNode, req *Reques
 	return dsNode, nil
 }
 
-// Execute runs the node and adds the results to vars. If the node requires
-// other nodes they must have already been executed and their results must
-// already by in vars.
-func (dn *DSNode) Execute(ctx context.Context, now time.Time, _ mathexp.Vars, s *Service) (r mathexp.Results, e error) {
-	logger := logger.FromContext(ctx).New("datasourceType", dn.datasource.Type, "queryRefId", dn.refID, "datasourceUid", dn.datasource.UID, "datasourceVersion", dn.datasource.Version)
-	ctx, span := s.tracer.Start(ctx, "SSE.ExecuteDatasourceQuery")
-	defer span.End()
-
-	pCtx, err := s.pCtxProvider.GetWithDataSource(ctx, dn.datasource.Type, dn.request.User, dn.datasource)
-	if err != nil {
-		return mathexp.Results{}, err
+// executeDSNodesGrouped groups datasource node queries by the datasource instance, and then sends them
+// in a single request with one or more queries to the datasource.
+func executeDSNodesGrouped(ctx context.Context, now time.Time, vars mathexp.Vars, s *Service, nodes []*DSNode) (e error) {
+	type dsKey struct {
+		uid   string // in theory I think this all I need for the key, but rather be safe
+		id    int64
+		orgID int64
+	}
+	byDS := make(map[dsKey][]*DSNode)
+	for _, node := range nodes {
+		k := dsKey{id: node.datasource.ID, uid: node.datasource.UID, orgID: node.orgID}
+		byDS[k] = append(byDS[k], node)
 	}
-	span.SetAttributes("datasource.type", dn.datasource.Type, attribute.Key("datasource.type").String(dn.datasource.Type))
-	span.SetAttributes("datasource.uid", dn.datasource.UID, attribute.Key("datasource.uid").String(dn.datasource.UID))
 
-	req := &backend.QueryDataRequest{
-		PluginContext: pCtx,
-		Queries: []backend.DataQuery{
-			{
+	for _, nodeGroup := range byDS {
+		firstNode := nodeGroup[0]
+		pCtx, err := s.pCtxProvider.GetWithDataSource(ctx, firstNode.datasource.Type, firstNode.request.User, firstNode.datasource)
+		if err != nil {
+			return err
+		}
+
+		logger := logger.FromContext(ctx).New("datasourceType", firstNode.datasource.Type,
+			"queryRefId", firstNode.refID,
+			"datasourceUid", firstNode.datasource.UID,
+			"datasourceVersion", firstNode.datasource.Version,
+		)
+
+		ctx, span := s.tracer.Start(ctx, "SSE.ExecuteDatasourceQuery")
+		defer span.End()
```

Comment:
Can we close the span after execute instead of using the defer? In for-loop it can lead to an unpredictable results and possible resource leaks. Also, it will make all spans to show correct end time.

Reply:
The are a few different place that return an error, so it looks like I would have to end the span manually in multiple places?

What about wrapping the loop body in a `func() { ... }()` for now -- will that fix the resource leaks? (Then in my error change PR, there will be less places to return to call the span end and I think I can remove it).

Reply:
I think extracting the loop's body in a function is a good way forward